azavea/raster-vision
# Add object detection integration test

## Overview

This PR adds an integration test that runs an object detection workflow. Similar to the classification test, it overfits a model on a synthetic scene, does eval on that same scene, and checks that the F1 score is 1.0. The test runs in ~5 minutes on my laptop and passes 5/5 times. If it fails we may need to increase the number of training steps.

This also fixes two bugs discovered along the way: #334 and #333

### Checklist


### Notes

There were a few unusual things that were required to train an object detection model in <= 5 mins and be able to memorize the data and get a perfect F1 score.
* We created a small test scene with colored squares using a synthetic scene generator to try to make it easier to learn. See `rastervision.contrib.scene_generator`.
* We added and used a new `sliding` option to `window_method` so that `make_training_chips` would use the same windows that the predict` command uses. This makes it  so the model would see the same images during training and prediction and have a chance to get to get an F1 of 1.0. Unfortunately, this doesn't exercise our usual windowing method.
* We turn off the eval and Tensorboard threads to conserve resources.
* We trained the model for 2500 steps to an F1 score of 0.9 and then use it as a pretrained model (stored at https://github.com/azavea/raster-vision-data/releases/tag/v0.0.5). This way, we give the model a head start, so it only needs to train for another 500 steps to get passed the finish line instead of the full 3000 steps.
* The default behavior of TFOD for finetuning models is to only copy the weights from the base of the model, but want to copy all weights from the pretrained model, so we had to set `load_all_detection_checkpoint_vars: true`. See https://github.com/tensorflow/models/issues/4944
* We turned off all data augmentation in `backend.config` so the model is trained on the same images it will see during prediction.

## Testing Instructions

* Change the `num_steps` in `backend.config` to 2501 and run `./scripts/integration_tests.py object-detection` locally. Check that it fails because the F1 score is < 1.0
* Check that this test runs in CI and that it passed.

Closes #315
Closes #259 
Closes #334
Closes #333
